Low-light Image Enhancement using MIRNet

soumik12345 opened the pull request:
With the goal of recovering high-quality image content from its degraded version, image restoration enjoys numerous applications, such as in surveillance, computational photography, medical imaging, and remote sensing. In this example, we implement the **MIRNet** model for low-light image enhancement, a fully convolutional architecture that learns an enriched set of features that combines contextual information from multiple scales, while simultaneously preserving the high-resolution spatial details.

Please find the corresponding notebook [here](https://colab.research.google.com/drive/1VtRwhq5sG_K3MNY6KYygtvUVC85Ju2qq?usp=sharing).

soumik12345 commented:
Hi @fchollet , can you please review this example?

fchollet commented on examples/vision/mirnet.py:
Please add a one-line description of this function

soumik12345 replied:
Cleaned up the inference code and removed the unnecessary `closest_number` function.

fchollet reviewed (commented):
Looking great, thanks for the update!

soumik12345 commented on examples/vision/mirnet.py:
For all the test images on LOLDataset, I added the corresponding enhanced images produced by a free image brightness correction tool from [pinetools](https://pinetools.com/change-image-brightness) in order to compare with the results achieved by MIRNet. Comparisons on the test images can be found in the [Colab Notebook](https://colab.research.google.com/drive/1VtRwhq5sG_K3MNY6KYygtvUVC85Ju2qq?usp=sharing).

soumik12345 commented on examples/vision/mirnet.py:
```diff
@@ -0,0 +1,516 @@
+"""
+## Creating a TensorFlow Dataset
+
+In order to restrict the training time to an hour, we would use 300 image pairs from the
```
Used 300 images for training, remaining 185 for validation.

soumik12345 commented on examples/vision/mirnet.py:
```diff
@@ -0,0 +1,516 @@
+- A feature extraction model that obtains a complementary set of features across multiple
+spatial scales, while maintaining the original high-resolution features to preserve
+precise spatial details.
+
```
Removed line breaks in between list items.

soumik12345 commented on examples/vision/mirnet.py:
```diff
@@ -0,0 +1,516 @@
+    )
+    gap = layers.GlobalAveragePooling2D()(combined_feature)
+    channel_wise_statistics = tf.reshape(gap, shape=(-1, 1, 1, channels))
+    compact_feature_representation = layers.ReLU()(
```
Passed `activation="relu"` to the Conv2D layer

soumik12345 commented on examples/vision/mirnet.py:
```diff
@@ -0,0 +1,516 @@
+            channel_wise_statistics
+        )
+    )
+    feature_descriptor_1 = tf.nn.softmax(
```
Passed `activation="softmax"` to the Conv2D layer

soumik12345 commented on examples/vision/mirnet.py:
Passed `activation="sigmoid"` to the Conv2D layer

soumik12345 commented:
Hi @fchollet , thanks for the suggestions, I have updated the example accordingly. Can you please review?

soumik12345 commented:
Updates:

- Added enhanced versions of the 15 test images from LOLDataset using an online tool called [pinetools](https://pinetools.com/change-image-brightness).
- Using 300 images for training, 185 for validation, and 15 as test images.
- Removed line breaks in between list items.
- Added `activation="relu"` in the Conv2D layers.
- Added `activation="softmax"` in the Conv2D layers.
- Added `activation="sigmoid"` in the Conv2D layers.
- Removed unnecessary functions in the inference.

fchollet commented on examples/vision/mirnet.py:
This is not a fair comparison, the enhancement here is extremely subpar. You can implement contrast normalization in pure Python or TensorFlow and it will work much better. Or just use PIL's `autocontrast`. The results will be much better and the comparison with the deep learning model will be fair.

soumik12345 replied:
Added comparison of the test images enhanced by MIRNet against the corresponding images enhanced using `PIL.ImageOps.autocontrast`.

fchollet commented on examples/vision/mirnet.py:
You can remove this section, inference is only valuable on test or validation sets

soumik12345 replied:
Removed inference on train images.

fchollet commented on examples/vision/mirnet.py:
You can also remove this section since you have a test section below

soumik12345 replied:
Removed inference on validation images.

soumik12345 commented:
Hi @fchollet , thanks for the suggestions. I have updated the example accordingly. Also updated the corresponding [Colab Notebook](https://colab.research.google.com/drive/1VtRwhq5sG_K3MNY6KYygtvUVC85Ju2qq?usp=sharing). Can you please review it?

- Added comparison of the test images enhanced by MIRNet against the corresponding images enhanced using `PIL.ImageOps.autocontrast`.
- Removed inference on train images.
- Removed inference on validation images.

fchollet reviewed (approved):
Excellent, it looks great! The comparison with `autocontrast()` does a great job at demonstrating the value of the MIRNet approach.

I pushed some copyedits, please pull them first. After pulling the copyedits, please add the generated files. I believe we're ready to publish 👍

soumik12345 commented:
Hi @fchollet , added the generated files.

fchollet commented:
Thanks for the great contribution! 👍
